### PR TITLE
fix: ILP with > 16 tags bug

### LINKED ISSUE #927: Is there a limit of 16 columns (tags) when dynamically creating a table using influx line protocol?
**Describe the bug**
When trying to write data using Influx Protocol, if I try to insert more than 16 tags (columns) the data is not inserted, yielding and exception, and questdb becomes instable (more details below).

**To Reproduce**

1. Bring the QuestDB up using this image:

```yaml
  version: "3.5"
  services:

    questdb:
      image: questdb/questdb:6.0.0-beta-linux-amd64
      restart: always
      ports:
        - 9000:9000
        - 9009:9009
        - 8812:8812
      network_mode: bridge

```

2. Run the following script in JS that will try to write this line:
Line:
`tableCRASH,tag_n_1=1,tag_n_2=2,tag_n_3=3,tag_n_4=4,tag_n_5=5,tag_n_6=6,tag_n_7=7,tag_n_8=8,tag_n_9=9,tag_n_10=10,tag_n_11=11,tag_n_12=12,tag_n_13=13,tag_n_14=14,tag_n_15=15,tag_n_16=16,tag_n_17=17 value=42.4 1619509249714000000`

Code:
```javascript
      ...
      const client = new net.Socket();
      client.connect(questDBInfluxPort, questDBHost, () => {
        let breakTags = '';
        for (let i = 1; i <= 17; i++) {
          breakTags += `tag_n_${i}=${i},`;
        }
        const r = `tableCRASH,${breakTags.substring(0, breakTags.length - 1)} value=42.4 ${Date.now() * 1e6}\n`;
        console.log('LINE WRITE QUEST_DB:', r);
        client.write(r);
        client.destroy();
      });

      client.on('data', function (data) {
        console.log('Received: ' + data);
      });

      client.on('close', function () {
        console.log('Connection closed');
      });
```
3. See error

```
...
questdb_1   | 2021-04-27T07:41:01.795739Z D i.q.c.SymbolMapReaderImpl open [name=/root/.questdb/db/tableCRASH/tag_n_15, fd=164, capacity=256]
questdb_1   | 2021-04-27T07:41:01.795794Z D i.q.c.v.SinglePageMappedReadOnlyPageMemory open /root/.questdb/db/tableCRASH/tag_n_16.o [fd=168, pageSize=64, size=64]
questdb_1   | 2021-04-27T07:41:01.795828Z D i.q.c.v.SinglePageMappedReadOnlyPageMemory open /root/.questdb/db/tableCRASH/tag_n_16.k [fd=169, pageSize=0, size=0]
questdb_1   | 2021-04-27T07:41:01.795933Z D i.q.c.v.SinglePageMappedReadOnlyPageMemory open /root/.questdb/db/tableCRASH/tag_n_16.v [fd=170, pageSize=0, size=0]
questdb_1   | 2021-04-27T07:41:01.796001Z D i.q.c.v.SinglePageMappedReadOnlyPageMemory open /root/.questdb/db/tableCRASH/tag_n_16.c [fd=171, pageSize=0, size=0]
questdb_1   | 2021-04-27T07:41:01.796068Z D i.q.c.SymbolMapReaderImpl open [name=/root/.questdb/db/tableCRASH/tag_n_16, fd=168, capacity=256]
questdb_1   | 2021-04-27T07:41:01.796347Z E i.q.c.l.t.LineTcpConnectionContext [35] could not process line data
questdb_1   | java.lang.ArrayIndexOutOfBoundsException: Index 16 out of bounds for length 16
questdb_1   | 	at io.questdb.std.IntList.getQuick(IntList.java:142)
questdb_1   | 	at io.questdb.cairo.TableReader.openSymbolMaps(TableReader.java:979)
questdb_1   | 	at io.questdb.cairo.TableReader.<init>(TableReader.java:109)
questdb_1   | 	at io.questdb.cairo.pool.ReaderPool$R.<init>(ReaderPool.java:353)
questdb_1   | 	at io.questdb.cairo.pool.ReaderPool.get(ReaderPool.java:85)
questdb_1   | 	at io.questdb.cairo.CairoEngine.getReader(CairoEngine.java:206)
questdb_1   | 	at io.questdb.cairo.CairoEngine.getReader(CairoEngine.java:198)
questdb_1   | 	at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$TableUpdateDetails$ThreadLocalDetails.getColumnIndex(LineTcpMeasurementScheduler.java:872)
questdb_1   | 	at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$LineTcpMeasurementEvent.createMeasurementEvent(LineTcpMeasurementScheduler.java:497)
questdb_1   | 	at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler.tryCommitNewEvent(LineTcpMeasurementScheduler.java:251)
questdb_1   | 	at io.questdb.cutlass.line.tcp.LineTcpConnectionContext.parseMeasurements(LineTcpConnectionContext.java:157)
questdb_1   | 	at io.questdb.cutlass.line.tcp.LineTcpConnectionContext.handleIO(LineTcpConnectionContext.java:147)
questdb_1   | 	at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$NetworkIOJobImpl.handleIO(LineTcpMeasurementScheduler.java:1152)
questdb_1   | 	at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$NetworkIOJobImpl.onRequest(LineTcpMeasurementScheduler.java:1115)
questdb_1   | 	at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:151)
questdb_1   | 	at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$NetworkIOJobImpl.run(LineTcpMeasurementScheduler.java:1135)
questdb_1   | 	at io.questdb.mp.Worker.run(Worker.java:112)
```

After exiting the NodeJS process, a kind of infinity loop happens and the log of that is:
```bash
...
questdb_1   | 2021-04-27T07:47:10.636816Z I i.q.c.l.t.LineTcpMeasurementScheduler releasing writer, its been idle since 2021-04-27T07:46:25.725000Z[tableName=tableCRASH]
questdb_1   | 2021-04-27T07:47:10.636829Z E server-main unhandled error [job=io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$WriterJob@6986bbaf, ex=
questdb_1   | java.lang.NullPointerException
questdb_1   | ]
questdb_1   | 2021-04-27T07:47:10.638222Z I i.q.c.l.t.LineTcpMeasurementScheduler releasing writer, its been idle since 2021-04-27T07:46:25.725000Z[tableName=tableCRASH]
questdb_1   | 2021-04-27T07:47:10.638235Z E server-main unhandled error [job=io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$WriterJob@6986bbaf, ex=
questdb_1   | java.lang.NullPointerException
questdb_1   | ]
questdb_1   | 2021-04-27T07:47:10.639604Z I i.q.c.l.t.LineTcpMeasurementScheduler releasing writer, its been idle since 2021-04-27T07:46:25.725000Z[tableName=tableCRASH]
questdb_1   | 2021-04-27T07:47:10.639619Z E server-main unhandled error [job=io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$WriterJob@6986bbaf, ex=
questdb_1   | java.lang.NullPointerException
questdb_1   | ]
questdb_1   | 2021-04-27T07:47:10.640948Z I i.q.c.l.t.LineTcpMeasurementScheduler releasing writer, its been idle since 2021-04-27T07:46:25.725000Z[tableName=tableCRASH]
questdb_1   | 2021-04-27T07:47:10.640962Z E server-main unhandled error [job=io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$WriterJob@6986bbaf, ex=
questdb_1   | java.lang.NullPointerException
...
```
**Expected behavior**
If there's a low column limit (16 is low IMHO), the documentation should state that upfront. If that's not the case, this write operation should have worked.

**Environment (please complete the following information):**
 - OS: Docker on MacOS
 - Browser: N/A
 - Version: 6.0.0-beta-linux-amd64 (but I also tried the last 5.x stable release and the problem also happens)

**Additional context**
If you run the same NodeJS code above but with **16** instead of **17** in the for loop, everything works fine. Of course, both tests, with 16 and 17 columns was made after stopping and removing the questdb container instance for a clean check.

### PR DESCRIPTION
**Pull request by ideoma**
Closes #927

### CODE REVIEW

**Review by patrickSpaceSurfer — APPROVED** [review]

File: core/src/main/java/io/questdb/cairo/TableReader.java [hunk 1]
```diff
@@ -95,6 +95,10 @@ public TableReader(CairoConfiguration configuration, CharSequence tableName) {
             LOG.info().$("table [id=").$(metadata.getId()).$(']').$();
             this.txFile = new TxReader(ff, path, partitionBy);
             readTxnSlow();
+            if (this.txn == TableUtils.INITIAL_TXN) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/TableReader.java:98]: this already done in txnFile. `readTxnSlow()` could have this:

```java
            // exit if this is the same as we already have
            if (txn == this.txn) {
                txnScoreboard.acquireTxn(txn);
                if(txn == TableUtils.INITIAL_TXN) {
                    this.txFile.readSymbolCounts(this.symbolCountSnapshot);
                }
                return false;
            }

```

File: core/src/main/java/io/questdb/mp/Worker.java [hunk 2]
```diff
@@ -152,13 +152,17 @@ public void run() {
...
 
     private void onError(int i, Throwable e) throws Throwable {
-        if (haltOnError) {
-            throw e;
-        }
-        if (log != null) {
-            log.error().$("unhandled error [job=").$(jobs.get(i).toString()).$(", ex=").$(e).$(']').$();
-        } else {
-            e.printStackTrace();
+        try {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/mp/Worker.java:155]: try... finally is excessive here, throwing out of finally block is not recommended

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 3]
```diff
@@ -2576,6 +2576,14 @@ private RecordCursorFactory generateTableQuery(
         }
     }
 
+    private static int getOrderByDirectionOrDefault(QueryModel model, int index) {
+        IntList direction = model.getOrderByDirectionAdvice();
+        if (index >= direction.size()) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:2581]: we can invert this for slight performance gain

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java (hunk 3 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:2581]: I believe this is default branch when order by is not specifed

File: core/src/main/java/io/questdb/cairo/TableReader.java (hunk 1 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/TableReader.java:98]: I changed it without adding to execution in readTxnSlow by setting initial txn to -1 which forces reader to read tx file on init

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**